willow/svg.py

```diff
@@ -0,0 +1,178 @@
+        if attr := (self.root.get("height") or self.root.get("width")):
+            return self._parse_size(attr)
+        elif self.view_box is not None:
+            return self.view_box.height
+        return 1
+
+    def _parse_size(self, raw_value):
+        clean_value = raw_value.strip()
+        match = self.UNIT_RE.search(clean_value)
+        unit = clean_value[match.start() :] if match else None
+
+        if unit == "%":
+            raise InvalidSizeAttribute(
+                f"Unable to handle relative size units ({raw_value})"
+            )
+
+        amount_raw = clean_value[: -len(unit)] if unit else clean_value
+        try:
+            amount = float(amount_raw)
+        except ValueError as err:
+            raise InvalidSizeAttribute(
+                f"Unable to parse value from '{raw_value}'"
+            ) from err
+        if amount <= 0:
+            raise InvalidSizeAttribute(f"Negative or 0 sizes are invalid ({amount})")
+
+        if unit is None or unit == "px":
+            return amount
+        elif unit == "em":
+            return amount * self.font_size_px
+        elif unit == "ex":
+            # This is not exactly correct, but it's the best we can do
+            return amount * self.font_size_px / 2
+        else:
+            return amount * self.dpi * self.COEFFICIENTS[unit]
+
+    def _get_view_box(self):
+        if attr := self.root.get("viewBox"):
+            return self._parse_view_box(attr)
+
+    def _parse_view_box(self, raw_value):
+        match = self.VIEW_BOX_RE.match(raw_value.strip())
+        if match is None:
+            raise ViewBoxParseError(f"Unable to parse viewBox value '{raw_value}'")
+        return ViewBox(*map(float, match.groups()))
+
+    def write(self, f):
+        self.dom.write(f, encoding="utf-8")
+
+
+class SVGImageTransform(str, Enum):
+    CROP = "crop"
+    RESIZE = "resize"
+
+
+class SVGImage(Image):
+    def __init__(self, image, operations=None):
+        self.image = image
+        self.operations = [] if operations is None else operations
+
+    @classmethod
+    @Image.converter_from(SVGImageFile)
+    def open(cls, svg_image_file):
+        svg_wrapper = SVGWrapper(svg_image_file.dom)
+        return cls(image=svg_wrapper)
+
+    @Image.operation
+    def get_size(self):
+        return (self.image.width, self.image.height)
+
+    @Image.operation
+    def auto_orient(self):
+        return self
+
+    @Image.operation
+    def crop(self, rect):
+        # Validate the rect here rather than when we actually apply
+        # the crop during rasterisation to give immediate feedback
+        left, top, right, bottom = rect
+        width, height = self.get_size()
+        if (
+            left >= right
+            or left >= width
+            or right <= 0
+            or top >= bottom
+            or top >= height
+            or bottom <= 0
+        ):
+            raise BadImageOperationError("Invalid crop dimensions: %r" % (rect,))
```

Comment:
This needs to be moved back into the svglib plugin - as the image may have been resized we need to test against the updated dimensions